### PR TITLE
Add business verification

### PR DESCRIPTION
**Pull request by sadick254**
Add support for business verification jobs.

### CODE REVIEW

File: src/web-api.ts [hunk 1]
```diff
@@ -600,33 +601,40 @@ export class WebApi {
...
+        validateReturnData(callbackUrl as string, data.return_job_status as boolean);
+      }
 
       if (jobType === 1) {
         validateEnrollWithId(image_details, (data.idInfo as IdInfo).entered);
       } else if (jobType === 6) {
         validateDocumentVerification(image_details);
       }
 
+      if (isBusinessVerification) {
```
Comment (mldangelo) [inline src/web-api.ts:634]: can you move this block to be above the `if (!isBusinessVerification) {` ?

File: test/web-api.test.ts [hunk 2]
```diff
@@ -678,6 +679,70 @@ describe('WebApi', () => {
...
+      };
+
+      nock('https://testapi.smileidentity.com')
+        .post('/v1/business_verification', (body) => {
+          expect(body.partner_id).toEqual('001');
+          expect(body.country).toEqual(id_info.country);
+          expect(body.id_type).toEqual(id_info.id_type);
+          expect(body.id_number).toEqual(id_info.id_number);
+          expect(body.business_type).toEqual(id_info.business_type);
+          expect(body.partner_params.job_type).toEqual(JOB_TYPE.BUSINESS_VERIFICATION);
```
Comment (mldangelo) [inline test/web-api.test.ts:704]: Prefer to use one large expect here where you can verify that you're only sending the keys you expect to send. 
```suggestion
          expect(body.partner_id).toEqual('001');
          expect(body.country).toEqual(id_info.country);
          expect(body.id_type).toEqual(id_info.id_type);
          expect(body.id_number).toEqual(id_info.id_number);
          expect(body.business_type).toEqual(id_info.business_type);
          expect(body.partner_params.job_type).toEqual(JOB_TYPE.BUSINESS_VERIFICATION);
```

Also consider moving this out of the post using a jest.fn so you can validate the expectation below.

File: test/web-api.test.ts [hunk 3]
```diff
@@ -678,6 +679,70 @@ describe('WebApi', () => {
...
+          return true;
+        })
+        .reply(200, business_verification_response.success)
+        .isDone();
```
Comment (mldangelo) [inline test/web-api.test.ts:708]: you should call an expectation on isDone() if you want to use it.

File: test/web-api.test.ts [hunk 4]
```diff
@@ -678,6 +679,70 @@ describe('WebApi', () => {
...
+
+      const resp = await instance.submit_job(partner_params, null, id_info, { signature: true });
+      expect(resp.data).toEqual(business_verification_response.success);
```
Comment (mldangelo) [inline test/web-api.test.ts:711]: Can you validate everything in the repsonse?

File: test/web-api.test.ts [hunk 5]
```diff
@@ -2,6 +2,7 @@ import path from 'path';
 import keypair from 'keypair';
 import nock from 'nock';
 import packageJson from '../package.json';
+import business_verification_response from './fixtures/business_verification_response.json';
```
Comment (mldangelo) [inline test/web-api.test.ts:5]: prefer camelCase here and below.

File: src/shared.ts [hunk 6]
```diff
@@ -23,6 +23,7 @@ export interface IdInfo {
   last_name?: string,
   dob?: string,
   phone_number?: string,
+  business_type?: string,
```
Comment (mldangelo) [inline src/shared.ts:26]: can you drop this in in sorted order?

**Review by mldangelo — CHANGES_REQUESTED** [review]

File: test/web-api.test.ts (hunk 4 above)
Comment (sadick254) [reply test/web-api.test.ts]: This does validate everything in the response. The `businessVerificationResp` contains two response `{ success: {/*all response*/}, unsupported_business_type: {/*all response*/}`

File: test/web-api.test.ts (hunk 3 above)
Comment (mldangelo) [reply test/web-api.test.ts:708]: https://github.com/nock/nock#isdone

File: examples/business_verification.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  user_id: '<put your unique ID for the user here>',
+  job_type: 1,
+};
+
+// Set fields required by the ID authority for a verification job.
+const id_info = {
+  country: '<2-letter country code>',
+  id_type: '<id type>',
+  id_number: '<valid id number>',
+  entered: 'true', // must be a string
```
Comment (mldangelo) [inline examples/business_verification.ts:28]: This can be a boolean now. See line 20 of `src/shared.ts`

**Review by mldangelo — DISMISSED** [review]
This is very close. Thank you. Please don't forget to review the changelog and readme.

**Review by tamssokari — DISMISSED** [review]
@sadick254 good job on this, one change though.

we want business verification to be in the `IDAPI` namespace for now and not `WebAPI`.
the decision framework we have adopted is that currently all text-only products are in the IDAPI namespace while those that make image uploads are in the `WebAPI` space.

bonus points for extracting `BusinessVerification` into it's own class and exposing that through IDAPI for now.

**Comment by ayinloya** [comment]
Looks great, same view as @tamssokari on calling business verification in idapi for now

File: examples/business_verification.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  job_type: 1,
+};
+
+// Set fields required by the ID authority for a verification job.
+const id_info = {
+  country: '<2-letter country code>',
+  id_type: '<id type>',
+  id_number: '<valid id number>',
+  entered: true,
+  busines_type: '<valid business type>', // this is optional except when country is NG
```
Comment (tamssokari) [inline examples/business_verification.ts:28]:
```suggestion
  business_type: '<valid business type>', // this is optional except when country is NG
  postal_address: '<valid postal address>', // this is optional, except when country is KE
  postal_code: '<valid postal address>', // this is optional, except when country is KE
```

**Review by tamssokari — COMMENTED** [review]
minor suggestion

**Review by mldangelo — COMMENTED** [review]
Before this goes in, look at how the idapi can be called from the webapi class. You should make a small modification there too.

File: src/id-api.ts [hunk 9]
```diff
@@ -65,6 +68,20 @@ export class IDApi {
         sid_server: this.sid_server,
       };
 
+      if (jobType === JOB_TYPE.BUSINESS_VERIFICATION) {
+        const signature = new Signature(this.partner_id, this.api_key)
+          .generate_signature(new Date().toISOString());
```
Comment (mldangelo) [inline src/id-api.ts:73]: nit, 

```suggestion
          .generate_signature();
```

File: src/web-api.ts [hunk 10]
```diff
@@ -593,13 +594,14 @@ export class WebApi {
     try {
       validatePartnerParams(partner_params);
 
-      if (parseInt(partner_params && partner_params.job_type.toString(), 10) === 5) {
+      const jobType = parseInt(partner_params.job_type.toString(), 10);
+      const useIDApi = jobType === JOB_TYPE.BASIC_KYC || jobType === JOB_TYPE.BUSINESS_VERIFICATION;
+      if (useIDApi) {
```
Comment (mldangelo) [inline src/web-api.ts:599]:
```suggestion
      const useIdApi = jobType === JOB_TYPE.BASIC_KYC || jobType === JOB_TYPE.BUSINESS_VERIFICATION;
      if (useIdApi) {
```

**Review by mldangelo — APPROVED** [review]
Thank you!

**Review by mldangelo — APPROVED** [review]

**Review by ayinloya — APPROVED** [review]

File: examples/business_verification.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+// Use '0' for the sandbox server, use '1' for production server.
+const sid_server = '<0 or 1>';
+
+const connection = new IDApi(partner_id, api_key, sid_server);
+
+// Create required tracking parameters
+const partner_params = {
+  job_id: '<put your unique job ID here>',
+  user_id: '<put your unique ID for the user here>',
+  job_type: 1,
```
Comment (ayinloya) [inline examples/business_verification.ts:19]: Can this be changed to 7 since it is business verification?
```suggestion
  job_type: 7,
```

**Review by mldangelo — APPROVED** [review]

**Pull request merged**